File: p381/p381b.go
// Project Euler Problem #381
/*
For a prime p let S(p) = ((p-k)!) mod(p) for 1 <= k <= 5.

For example, if p=7,
(7-1)! + (7-2)! + (7-3)! + (7-4)! + (7-5)! = 6! + 5! + 4! + 3! + 2! = 720+120+24+6+2 = 872.
As 872 mod(7) = 4, S(7) = 4.
It can be verified that Sum(S(p)) = 480 for 5 <= p < 100.
Find Sum(S(p)) for 5 <=  p < 10^8.
*/

package main

import (
    "fmt"
	"math"
    "math/big"
)

func isPrime(n int) bool {
	num := float64(n)
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n % 2 ==0 { 
	 return false
	}
	x := 3
	limit := int(math.Sqrt(num))
	for x <= limit {
		if n % x == 0 {
			return false
		}
		x += 2
	}
	return true
}

func Factorial(x float64) float64 {
    if x == 0 {
        return 1;
    } 
    return x * Factorial(x-1);
}

func S(p float64) float64 {  
    var sum float64 = 0
    var k float64    
    for k = 1; k <6; k++ { 
//        fmt.Println(p,k,p-k,Factorial(p-k))
        sum += Factorial(p-k)
    }
    return math.Mod(sum,p)
}

func SumOfS(i,j float64) float64 {
    var sum float64 = 0
    for p:=i; p < j; p++ {
        if isPrime(int(p)) {
//            fmt.Println(p, S(p))
            sum+= S(p)
        }
    }
    return sum  
}

func main() {
    if S(7) == 4 {
        fmt.Println("Test #1 passed.")
    } else {
        fmt.Println("Test #1 failed!")
    }  
   

    if SumOfS(5,100) == 480 {
        fmt.Println("Test #2 passed.")
    } else {
        fmt.Println("Test #2 failed!", SumOfS(5,100) )
    }        
}
//        
//    result = sum([S(p) for p in xrange(5,100000000,2) if isPrime(p,primeSet, primeList)])  
//    prfloat64 "The answer to #381 is: %s" %result 

File: p381/p381b.py
#!/usr/bin/env python
# Project Euler Problem #381

'''
For a prime p let S(p) = ((p-k)!) mod(p) for 1 <= k <= 5.

For example, if p=7,
(7-1)! + (7-2)! + (7-3)! + (7-4)! + (7-5)! = 6! + 5! + 4! + 3! + 2! = 720+120+24+6+2 = 872.
As 872 mod(7) = 4, S(7) = 4.
It can be verified that Sum(S(p)) = 480 for 5 <= p < 100.
Find Sum(S(p)) for 5 <=  p < 108.
'''
from math import factorial as fac, sqrt

primeSet  = set([2,3])
primeList = [2,3]   

def isPrime(n,primeSet = None, primeList = None):
    if primeSet == None:
        primeSet  = set([2,3])
        primeList = [2,3]  

    if n in primeSet:
        return True
    elif n < 2:
        return False    
    else:
        limit = int(sqrt(n))+1    
        for x in primeList: 
            if x >= limit:
                break
            else:
                if n % x == 0:
                    return False
        primeSet.add(n)
        primeList.append(n)
        return True

def sundaram(max_n):
    numbers = range(3, max_n+1, 2)
    half = (max_n)//2
    initial = 4 # was 4
    for step in xrange(3, max_n+1, 2):
        for i in xrange(initial, half, step):
            numbers[i-1] = 0
        initial += 2*(step+1)

        if initial > half:
            return [2] + filter(None, numbers)

def S(p):
    value = 0
    #(p-3) term
    value+= (p-1)/2
    #(p-4) term
    if (p+1)%6 ==0:
        value += (p+1)/6
    else:
        value += (5*p+1)/6
    #(p-5) term
    if (p-1)%24 == 0:
        value += (p-1)/24
    else:
        value += ((p**2-1)/24)%p
    return value % p
           
if __name__ == "__main__":
    #verify test problem
    if S(7) == 4:
        print "test #1 passed."
    else:
        print "test failed!", S(7) 
    
    if sum([S(p) for p in range (5,100) if isPrime(p,primeSet, primeList)]) == 480:    
        print "test #2 passed."
    else:
        print "test #2 failed!"     
        
    result = sum([S(p) for p in sundaram(10**8) if p not in [2,3]])  
    print "The answer to #381 is: %s" %result 
